paddle/fluid/operators/center_loss_op.cu

```diff
@@ -0,0 +1,133 @@
+namespace paddle {
+namespace operators {
+
+using platform::PADDLE_CUDA_NUM_THREADS;
+
+template <typename T>
+__global__ void ComputeDistance(int64_t total_num, const int K, const T *X,
+                                const int64_t *label, const T *centers_data,
+                                T *distance) {
+  for (int index = blockIdx.x * blockDim.x + threadIdx.x; index < total_num;
+       index += blockDim.x * gridDim.x) {
+    int m = index / K;
+    int k = index % K;
+    int64_t label_value = label[m];
+    distance[index] = X[index] - centers_data[label_value * K + k];
+  }
+}
+
+template <typename T>
+__global__ void ComputeCenterDiffGpu(int num, const int M, const int K,
+                                     const int64_t *label, const T *distance,
+                                     T *variation_sum, const T *alpha,
+                                     T *centers_out) {
+  for (int64_t index = blockIdx.x * blockDim.x + threadIdx.x; index < num;
+       index += blockDim.x * gridDim.x) {
+    int count = 0;
+    for (int m = 0; m < M; m++) {
+      int64_t label_value = label[m];
+      if (label_value == index) {
+        count++;
+        for (int k = 0; k < K; k++) {
+          variation_sum[index * K + k] += distance[m * K + k];
+        }
+      }
+    }
+    for (int k = 0; k < K; k++) {
+      centers_out[index * K + k] +=
+          alpha[0] * variation_sum[index * K + k] / (count + (T)1.);
+    }
+  }
+}
+
+template <typename DeviceContext, typename T>
+class CenterLossCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto &device_context = ctx.template device_context<DeviceContext>();
+    auto stream = device_context.stream();
+    auto *X = ctx.Input<Tensor>("X");  // deep feature
+    auto *labels = ctx.Input<Tensor>("Label");
+    auto *centers = ctx.Input<Tensor>("Centers");
+    auto *update_rate = ctx.Input<Tensor>("CenterUpdateRate");
+    int cluster_num = ctx.Attr<int>("cluster_num");
+    auto *lr_center = update_rate->data<T>();
+    bool need_update = static_cast<T>(ctx.Attr<bool>("need_update"));
+
+    auto x_data = X->data<T>();
+    auto label_data = labels->data<int64_t>();
+
+    auto x_dims = X->dims();
+    int batch_size = x_dims[0];
+    const int deep_feat_dim = x_dims[1];
+
+    auto *centers_diff = ctx.Output<Tensor>("CentersDiff");
+    auto centers_diff_data = centers_diff->mutable_data<T>(ctx.GetPlace());
+
+    auto centers_data = centers->data<T>();
+    auto centers_dim = centers->dims();
+    auto *out_loss = ctx.Output<Tensor>("Loss");
+    auto loss_data = out_loss->mutable_data<T>(ctx.GetPlace());
+
+    auto *centers_out = ctx.Output<Tensor>("CentersOut");
+    auto *centers_out_data = centers_out->mutable_data<T>(ctx.GetPlace());
+
+    if (centers_out_data != centers_data) {
+      int size = centers_out->numel() * sizeof(T);
+      cudaMemcpyAsync(centers_out_data, centers_data, size,
```

Comment:
I think this should be equivalent with cudaMemcpy, since you only use a single stream here.

Reply:
suggest TensorCopy high level api